### PR TITLE
New AddTakeGradLargeBatch and SortByKey for GPUs

### PR DESCRIPTION
**Pull request by ap-hynninen**

### CODE REVIEW

File: src/operator/tensor/indexing_op.h [hunk 1]
```diff
@@ -312,15 +407,7 @@ void TakeOpBackward(const nnvm::NodeAttrs& attrs,
             if ((grad_out.shape_[0] < grad_out.shape_[1]) && (grad_out.shape_[0] < 512)) {
```
Comment (piiswrong) [inline src/operator/tensor/indexing_op.h]: Since largebatch is faster now maybe change 512 to a smaller number?

Comment (ap-hynninen) [inline src/operator/tensor/indexing_op.h]: Sure. I'll go back and see what is optimal.

Comment (ap-hynninen) [inline src/operator/tensor/indexing_op.h]: unsigned long long shape_in_prod = 
        static_cast<unsigned long long>(grad_in.shape_[0])*
        static_cast<unsigned long long>(grad_in.shape_[1]);

      unsigned long long shape_out_prod = 
        static_cast<unsigned long long>(grad_out.shape_[0])*
        static_cast<unsigned long long>(grad_out.shape_[1]);

      if (shape_out_prod < 16384ULL && shape_in_prod < 16384ULL) {
        AddTakeGrad(grad_in, data, grad_out);

This is a new criteria I came up with. Should I put it in a separate PR?

Comment (piiswrong) [inline src/operator/tensor/indexing_op.h]: you can update this one.

Comment (piiswrong) [inline src/operator/tensor/indexing_op.h]: btw, use uint64_t

Comment (piiswrong) [inline src/operator/tensor/indexing_op.h]: and could you put some comment explaining the magic number?

**Comment by piiswrong** [comment]
LGTM
please fix lint and we can merge

File: src/operator/tensor/sort_op.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+ * \param num_keys number of keys to sort
+ */
+template <typename KDType, typename VDType, typename xpu>
+inline typename std::enable_if<std::is_same<xpu, gpu>::value, size_t>::type
+SortByKeyWorkspaceSize(const size_t num_keys);
+
+}  // namespace op
+}  // namespace mxnet
+#ifdef __CUDACC__
+#include "./sort_op-inl.cuh"
```
Comment (piiswrong) [inline src/operator/tensor/sort_op.h]: Does it work if we include this only in .cu files?

File: src/operator/tensor/indexing_op.h [hunk 3]
```diff
@@ -375,4 +482,7 @@ void BatchTakeOpForward(const nnvm::NodeAttrs& attrs,
 
 }  // namespace op
 }  // namespace mxnet
+#ifdef __CUDACC__
+#include "./indexing_op-inl.cuh"
```
Comment (piiswrong) [inline src/operator/tensor/indexing_op.h]: same

File: src/operator/tensor/indexing_op-inl.cuh [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+  int iidx_end = (idx_start == NULL) ? ymax : *idx_start_size_ptr;
+
+  for (int iidx = blockIdx.y;iidx < iidx_end;iidx += gridDim.y) {
+
+    // Thread block sums up elements in the range [idx_begin, idx_end-1]
+    int idx_begin, idx_end;
+    int sorted_value;
+    if (idx_start == NULL) {
+      idx_begin = iidx;
+      sorted_value = static_cast<int>(sorted[idx_begin]);
```
Comment (sxjscience) [inline src/operator/tensor/indexing_op-inl.cuh]: I find that if the IdxType is float, directly casting may be problematic for values like 9.99999. Should we try to solve it? The previous code has also not handled this type of special case.

Comment (ap-hynninen) [inline src/operator/tensor/indexing_op-inl.cuh]: I'd suggest that we solve this in a separate PR

Comment (ap-hynninen) [inline src/operator/tensor/indexing_op-inl.cuh]: I think it would be best to round() all float index arrays to int early on.

Comment (sxjscience) [inline src/operator/tensor/indexing_op-inl.cuh]: @ap-hynninen I see. round in the early stage should be more reasonable.

**Pull request merged**